src/ResourceManager/Common/Commands.Common.Strategies/AsyncCmdletExtensions.cs

```diff
@@ -0,0 +1,206 @@
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    public static class AsyncCmdletExtensions
+    {
+        /// <summary>
+        /// WriteVerbose function with formatting.
+        /// </summary>
+        /// <param name="cmdlet">Cmdlet</param>
+        /// <param name="message">message with formatting</param>
+        /// <param name="p">message parameters</param>
+        public static void WriteVerbose(this IAsyncCmdlet cmdlet, string message, params object[] p)
+            => cmdlet.WriteVerbose(string.Format(message, p));
+
+        /// <summary>
+        /// The function read current Azure state and update it according to the `parameters`.
+        /// </summary>
+        /// <typeparam name="TModel">A resource model type.</typeparam>
+        /// <param name="client">Azure SDK client.</param>
+        /// <param name="subscriptionId">Subbscription Id.</param>
+        /// <param name="parameters">Cmdlet parameters.</param>
+        /// <param name="asyncCmdlet">Asynchronous cmdlet interface.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns></returns>
+        public static async Task<TModel> RunAsync<TModel>(
+            this IClient client,
+            string subscriptionId,
+            IParameters<TModel> parameters,
+            IAsyncCmdlet asyncCmdlet,
+            CancellationToken cancellationToken)
+            where TModel : class
+        {
+            // create a DAG of configs.
+            var config = await parameters.CreateConfigAsync();
+            // read current Azure state.
+            var current = await config.GetStateAsync(client, cancellationToken);
+            // update location.
+            parameters.Location = current.UpdateLocation(parameters.Location, config);
+            // update a DAG of configs.
+            config = await parameters.CreateConfigAsync();
+            // create a target.
+            var target = config.GetTargetState(
+                current, new SdkEngine(subscriptionId), parameters.Location);
+            // print paramaters to a verbose stream.
+            foreach (var p in asyncCmdlet.Parameters)
+            {
+                asyncCmdlet.WriteVerbose(p.Key + " = " + ToPowerShellString(p.Value));
+            }
+
+            // apply the target state
+            var newState = await config.UpdateStateAsync(
+                client,
+                target,
+                cancellationToken,
+                new ShouldProcess(asyncCmdlet),
+                asyncCmdlet.ReportTaskProgress);
+            // return a resource model
+            return newState.Get(config) ?? current.Get(config);
+        }
+
+        static string UpdateLocation(
```

Comment:
I don't think that the default should be east us, I think this is something that should be set for each client (perhaps a DefaultLocation property.

The fear is that services that are not available in eastus will default to this value inadvertently